### ISSUE TITLE
Audit and evaluate apiserver client usage in kubelet

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#14702]
The current apiserver client QPS for kubelet is the default 5.0, which is not high for kubelet. When batch creating pods, the startup latency is completely dominated by the QPS (e.g. see https://github.com/kubernetes/kubernetes/issues/14554#issuecomment-143829191, events consumed most of the bandwidth). Moreover, the *perceived* kubelet performance may be affected by slight change in kubelet's behavior, due to the QPS limitation (e.g., https://github.com/kubernetes/kubernetes/issues/14554#issuecomment-143898216).

We should audit our apiserver client usage and re-evaluate whether to bump the QPS, or to clearly surface any increase in the client usage.

Major consumers for the apiserver client includes:
 * events
 * pod status
 * node status (relatively low)

The breakdown of starting a pod of N containers (when there is no failure):
 * (N+1) * 3 events  (for each container, including the infra container, there are pulled, created, and started events)
 * 3 pod status update (the initial pending, running, and finally the running+ready status). This may increase once we let user control per-container probing period.

For starting 30 single-container pods on kubelet, the pods along would generate ((1+1)*3+3)*30= 9*30=270 requests.
This would take kubelet 270/5 qps = 54 seconds to send out.

/cc @kubernetes/goog-node

**Comment by wojtek-t** [comment]
I agree events are very problematic (the aspect you mentioned in this bug is one of those). I think that we should finally get to moving events to separate storage (either another etcd instance or some other storage) and then we can probably have separate QPS limits.

**Comment by yujuhong** [comment]
> I agree events are very problematic (the aspect you mentioned in this bug is one of those). I think that we should finally get to moving events to separate storage (either another etcd instance or some other storage) and then we can probably have separate QPS limits.

Has someone already started working on moving the events to a separate storage? That'd be great :)
Kubelet does support separate QPS limits for events (which uses a different apiserver client), but it's currently not used. I tried using this yesterday and set a low 1.0 QPS limit for events, and it didn't turn out as I expected. The throttled events slowed down the entire kubelet. I didn't have time to check whether we could solve this by adjusting the event queue length, etc. It might be worth exploring this as a short-term solution.

What remains is that we'd still need to keep an eye on the various status updates vs the QPS limit.

**Comment by wojtek-t** [comment]
@yujuhong - yes - I'm looking into it - I have an extremely hacky prototype here:
https://github.com/kubernetes/kubernetes/compare/master...wojtek-t:separate_events_etcd
that I used to verify performance impact of it. It seems to be pretty significant in large cluster, so I will be working on making it reviewable and mergable this week.

**Comment by yujuhong** [comment]
@wojtek-t, thanks for the update :)

**Comment by yujuhong** [comment]
For measurement, I think we can instrument and/or stub out the rest client in the apiserver client to monitor the usage, similar to what we do with Docker.

**Comment by jiangyaoguo** [comment]
+1 for separate client for events. Do we should turn down original 5 QPS to keep the overall QPS of kubelet?

**Comment by jiangyaoguo** [comment]
+1 for separate client for events. Do we should  keep overall QPS of kubelet in view of two separate apiserver client?

发件人: Yu-Ju Hong [mailto:notifications@github.com]
发送时间: 2015年11月17日 4:05
收件人: kubernetes/kubernetes
抄送: Jiangyaoguo
主题: Re: [kubernetes] Audit and evaluate apiserver client usage in kubelet (#14702)


@wojtek-t<https://github.com/wojtek-t>, now that events have been moved to a separate etcd instance, we can set -event-qps so that kubelet will create a separate (apiserver) client for events<https://github.com/kubernetes/kubernetes/blob/4ad3d6f5fe6e9e4c7067a2e3f76735d2d8bf1e28/cmd/kubelet/app/server.go#L779>. Do you have any suggestions on the numbers to use?
Also, it seems to me that there is no reason not to always use a separate client for events now. WDYT?

**Comment by wojtek-t** [comment]
>
@wojtek-t, now that events have been moved to a separate etcd instance, we can set -event-qps so that kubelet will create a separate (apiserver) client for events. Do you have any suggestions on the numbers to use?
Also, it seems to me that there is no reason not to always use a separate client for events now. WDYT?

It definitely makes sense to have a separate client for events.
However, I'm not sure what the limit should be. I think the right way to proceed is to add support for a separate client and once this is done experiment with different values of the limit and check how this is affecting the overall system performance.

**Comment by yujuhong** [comment]
Since this may affect the cluster scalability, you and @gmarek may have a better idea on how to test this. Do we have existing clusters/benchmarks for this purpose?

**Comment by wojtek-t** [comment]
@yujuhong - once this is implemented, I'm happy to do some benchmarks to get values for limits

**Comment by gmarek** [comment]
+/infty for this effort. If your change will be compatible with SimpleKubelet, you can easily modify HollowKubelet code and run Kubemark experiments with your change, if you don't have access to 'big' projects. Default quota is enough to run 100 'Node' Kubemark.

**Comment by yujuhong** [comment]
/remove-lifecycle stale

Issue still seem relevant.
